proposals/0000-hyperdb.md

```diff
@@ -0,0 +1,581 @@
+prefix does not exist, returns an empty list. The order of returned keys is
+implementation (or configuration) specific. Read-only.
+
+If the hypercore underlying a hyperdb is only partially replicated, behavior is
+implementation-specific. For example, a `get()` call could block until the
+relevant value is replicated, or the implementation could return an error.
+
+An example pseudo-code session working with a database might be:
+
+    db.put('/life/animal/mammal/kitten', '{"cuteness": 500.3}')
+    db.put('/life/plant/bush/bananna', '{"delicious": 103.4}')
+    db.delete('/life/plant/bush/bananna')
+    db.put('/life/plant/tree/bananna', '{"delicious": 103.4}')
+    db.get('/life/animal/mammal/kitten')
+    => {"cuteness": 500.3}
+    db.list('/life/')
+    => ['/life/animal/mammal/kitten', '/life/plant/tree/bananna']
+
+
+# Reference Documentation
+[reference-documentation]: #reference-documentation
+
+A HyperDB hypercore feed typically consists of a sequence of entries, all of
+which are protobuf-encoded Node messages. Higher-level protocols may make
+exception to this: for example, hyperdrive reserves the first entry of the
+(`metadata`) feed for a special entry that refers to the second (`content`)
+feed.
+
+The sequence of entries includes an incremental index: the most recent entry in
+the feed contains metadata pointers that can be followed to efficiently look up
+any key in the database without needing to linear scan the entire history or
+generate an independent index data structure. Of course implementations are
+free to maintain such an index if they prefer.
+
+The Node protobuf message schema is:
+
+```
+  message Node {
+    optional string key = 1;
+    optional bytes value = 2;
+    repeated uint64 clock = 3;
+    optional bytes trie = 4;    // TODO: actually a {feed, seq} pair
+    optional bytes path = 5;    // TODO: actual type?
+    optional uint64 seq = 6;
+    optional bytes feed = 7;    // TODO: actual type?
+  }
+```
+
+TODO(mafintosh): where is this schema actually defined for the `next` branch of
+the hyperdb repo?
+
+- `key`: UTF-8 key that this node describes. Leading and trailing forward
+  slashes (`/`) are always striped before storing in protobuf.
+- `value`: arbitrary byte array. A non-existant `value` entry indicates that
+  this Node indicates a deletion of the key; this is distinct from specifying
+  an empty (zero-length) value.
+- `clock`: reserved for use in the forthcoming `multi-writer` standard, not
+  described here. An empty list is the safe (and expected) value for `clock` in
+  single-writer use cases.
+- `trie`: a structured array of pointers to other Node entries in the feed,
+  used for navigating the tree of keys.
+- `path`: a 2-bit hash sequence of `key`'s components.
+- `seq`: the zero-indexed sequence number of this Node (hyperdb) entry in
+  the feed. Note that this may be offset from the hypercore entry index if
+  there are any leading non-hyperdb entries.
```

Comment:
Does this mean that `seq` only increments when a hyperdb Node is written to the feed?